docs/proposals/THV-2106-virtual-mcp-server.md

```diff
@@ -0,0 +1,1035 @@
+
+      - id: "create_report"
+        tool: "jira.create_issue"
+        arguments:
+          title: "Incident {{.params.incident_id}} Analysis"
+          description: |
+            Application Logs: {{.steps.fetch_app_logs.output}}
+            Error Metrics: {{.steps.fetch_error_metrics.output}}
+            Trace Data: {{.steps.fetch_trace_data.output}}
+            Infrastructure: {{.steps.fetch_infra_status.output}}
+```
+
+**Note**: The `investigate_incident` composite tool demonstrates calling the same backend tool (`fetch`) multiple times with different static URLs to gather incident data from various monitoring sources. This pattern allows a single generic tool to be reused with different static endpoints, creating a purpose-built workflow without needing separate tools for each data source.
+
+**CLI Usage:**
+
+```bash
+# Start Virtual MCP from config file
+thv virtual --config virtual-mcp-config.yaml
+
+# With environment variables
+export VMCP_CLIENT_SECRET="..."
+export GITHUB_EXCHANGE_SECRET="..."
+export JIRA_EXCHANGE_SECRET="..."
+export SLACK_BOT_TOKEN="..."
+thv virtual --config virtual-mcp-config.yaml
+
+# Override specific settings
+thv virtual --config virtual-mcp-config.yaml \
+  --port 8080 \
+  --log-level debug
+```
+
+---
+
+### Kubernetes Configuration
+
+**New CRD: `VirtualMCPServer`**
+
+```yaml
+apiVersion: toolhive.stacklok.dev/v1alpha1
+kind: VirtualMCPServer
+metadata:
+  name: engineering-vmcp
+  namespace: default
+spec:
+  # Reference to ToolHive group
+  groupRef:
+    name: engineering-team
+
+  # ===== INCOMING AUTHENTICATION (Client → Virtual MCP) =====
+  incomingAuth:
+    # Virtual MCP's own OIDC configuration
+    oidcConfig:
+      type: kubernetes  # kubernetes | inline | configMapRef
+      kubernetes:
+        audience: vmcp
+        # Uses default Kubernetes OIDC discovery
+
+    # Optional: Authorization policies
+    authzConfig:
+      type: inline
+      inline:
+        policies:
+          - |
+            permit(
+              principal,
+              action == Action::"tools/call",
+              resource
+            );
+
+  # ===== OUTGOING AUTHENTICATION (Virtual MCP → Backends) =====
+  outgoingAuth:
+    # Configuration source
+    source: discovered  # inline | discovered | mixed
+
+    # When source=discovered:
+    # Virtual MCP queries each backend's MCPServer.spec.externalAuthConfigRef
+    # and uses those MCPExternalAuthConfig resources for token exchange
+
+    # Optional: Default behavior for backends without auth config
+    default:
+      type: pass_through
+
+    # Optional: Override specific backends (works in all modes)
+    backends:
+      slack:
+        type: service_account
+        serviceAccount:
+          credentialsRef:
+            name: slack-bot-token
+            key: token
+          headerName: Authorization
+          headerFormat: "Bearer {token}"
+
+  # ===== TOKEN CACHING =====
+  tokenCache:
```

Comment:
This could be reusable if it's redis, wondering if we could have a custom CRD for TokenCaching that can be referenced via `tokenCacheRef`. Saves repeating redis/equivalent connection strings